use bitfield to shrink TensorImpl

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45263 use bitfield to shrink TensorImpl**
* #45046 faster TensorOptions merging
* #45043 track Half/ComplexHalf default dtype
* #45026 pass TypeMeta by value

Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

Commit:
use bitfield to shrink TensorImpl

[ghstack-poisoned]

ezyang commented:
@glaringlee ISTR you were working on a patchset for this, whatever happened to it?

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -48,10 +48,14 @@ TensorImpl::TensorImpl(
     const caffe2::TypeMeta data_type)
-    : TensorImpl(std::move(storage), key_set, data_type, storage.device()) {}
+    : TensorImpl(std::move(storage), key_set, data_type, storage.device()) {
+  init_bitfields();
```
These seem unnecessary, if we call the TensorImpl constructor that should take care of initialization, no?

bhosmer replied:
Oops yes, thanks for catching

glaringlee commented:
@ezyang Ah, it seems `:1` is preferred here.

I do have my impl on my local machine, but haven't send PR since I was planing on testing the perf between `:1` on each `bool` (except the is_contiguous one) vs `using a uint16_t` to pack all of them. Then switched on C++ Frontend work to catch up 1.7 release. 

`:1` has no overhead, but currently we have 8 bools to pack, so 1 byte is enough. if there is one more flag needed in the future, then another bitfield means another byte allocation (correct me if I am wrong, and I am not sure whether 29 int64_t can still hold one more byte). So I was thinking of using a uint16_t to pack all these booleans and then we still have room for more flags. But this will introduce as least 1 more instruction (xor?) when retrieving each bit. So i have both impl there, but did not test the pref yet :).

ezyang commented:
I mean, `: 1` just means the compiler is going to generate the masking, rather than you doing it manually. It does seem like it's a lot easier to write though.

glaringlee commented:
> I mean, `: 1` just means the compiler is going to generate the masking, rather than you doing it manually. It does seem like it's a lot easier to write though.


Right, compiler will generate masking for `:1`. I have no problem on using `:1`. 

Here is what I've been thinking. A bit long though. 
Currently we have 9 booleans in ,  `bool is_contiguous_` still uses a byte and the rest of 8 booleans are generated into a 1 byte bit mask by using `:1`. This is fine.

I was trying to use an uint16_t since uint16_t has 16 bits, so we still have room for newly added flag if any (7 free slots). If using `:1`, then if a new boolean flag is added into TensorImpl, then we need one more bit, but since the memory alignment is byte level, so I think adding a new boolean flag will cause one byte more memory allocation anyway even putting `:1` there. (since the last byte is already full with 8 bits flags). The size limit of TensorImpl is changed to 29 int64_t, I don't know whether this 29 int64_t have one more byte free space, if not, adding a new bool flag will cause the size to 30 int64_t. On the other hand, if using uint16_t to pack all the boolean flags, then retrieving each bit will need an xor instruction, which will slow down the perf, I am not sure whether compiler will optimize this,  that's the perf thing I was planing to test see whether uint16_t is good enough to use. (Please correct me if I am wrong anywhere).

But yes, if adding new flag, we can turn is_contiguous_ using bit as well, so the total size of tensor won't change.  So `:1` is good to go.

bhosmer commented:
> But yes, if adding new flag, we can turn is_contiguous_ using bit as well, so the total size of tensor won't change. So `:1` is good to go.

@glaringlee yup I think that's right - as soon as we go to > 9 bools we can just start a second bitfield, with `is_contiguous_` and its new friends. 

BTW if I'm calculating right, this PR makes the layout look like this
```
               offset   size
...
data_type_        208      2
device_opt_       210      6
key_set_          216      8
is_contiguous_    224      1 // future home of bitfield 2
bitfield          225      1 // is_channels_last_ ... reserved_
<empty>           226      6
                  232
```
...so yeah, we're actually almost at 224 already - e.g. either of the two approaches to `device_opt_` in the description would do it:
```
               offset   size
...
data_type_        208      2
device_           210      4
key_set_          214      8
bitfield          222      1 // is_contiguous_, has_device_, ...
bitfield          223      1 // is_channels_last_ ... reserved_
                  224

               offset   size
...
data_type_        208      2
device_           210      4 // use !undefined as proxy for has_device_
key_set_          214      8
is_contiguous_    222      1
bitfield          223      1 // is_channels_last_ ... reserved_
                  224
```

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

swolchok commented:
I wonder if the flags could be coalesced better. For example, `is_channels_last_` and `is_channels_last_3d_` are mutually exclusive. Perhaps there's a relationship between the other `is_channels_foo` `bool`s as well?

bhosmer commented:
> I wonder if the flags could be coalesced better. For example, `is_channels_last_` and `is_channels_last_3d_` are mutually exclusive. Perhaps there's a relationship between the other `is_channels_foo` `bool`s as well?

@swolchok possibly! Though it's out of scope for this PR, as my aim was just to compress the structure without changing any of the semantics (e.g. that's why this PR doesn't include the `device_opt_` optimization  mentioned in the description, which would save another 8 bytes). 

But yeah, I wouldn't be surprised if there are other opportunities here. I'm not an expert in the channels stuff, but cc @VitalyFedyunin (just based on blaming the relevant lines)

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]

smessmer commented on c10/core/TensorImpl.h:
```diff
@@ -1675,36 +1675,47 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   bool is_contiguous_ = true;
 
+  // default member initializers for bit-fields only available with -std=c++2a or -std=gnu++2a
+  inline void init_bitfields() {
```
nit: no need for `inline` keyword when you're in a class body

smessmer commented on c10/core/TensorImpl.h:
```diff
@@ -1675,36 +1675,47 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   // Channels last contiguous tensor is channel last tensor which occupies
   // contiguous memory block.
-  bool is_channels_last_contiguous_ = false;
+  bool is_channels_last_contiguous_ : 1;
```
keeping default values here didn't work?

bhosmer replied:
nah, bitfields don't allow an initializer as it turns out

Commit:
Update on "use bitfield to shrink TensorImpl"


Packing 8 (of the 9) bools at the end of `TensorImpl` takes it from 240 bytes down to 232. Bump in instruction count is minor, interested to see what in-house microbenchmarks say. 

(BTW no idea if there's a reason this hasn't been done already: [a comment](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1674-L1675) seems to imply it was intended to be done at some point, but I'm putting this up in complete ignorance of any other history behind the current state.)

Possible next step: replace `device_opt_` with `device_` and `has_device_`, a la `TensorOptions`, and pack `has_device_` and `is_contiguous_` together. That would get us down to 224. 

Alternative: [a code comment describes an invariant](https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1663-L1664) which IIUC means that the optionality of `device_opt_` moves in lockstep with whether a tensor is defined or not. If so, the optionality is superfluous, and we can get to 224 without another bitfield (and be correct by construction in the bargain), simply by using (tensor is not undefined) as our `has_device_`. 

Differential Revision: [D23900587](https://our.internmc.facebook.com/intern/diff/D23900587)

[ghstack-poisoned]
